File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableValueCodec.cs
using System.Collections.Immutable;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>The outcome of projecting a presented value onto the version a consumer recognizes.</summary>
public sealed record PortableProjection(PortableValue Value, ImmutableArray<string> MappingObligations);

/// <summary>
/// The schema-guided value codec: the negotiated Shape determines the type of every position, so
/// values carry no kind discriminator on the wire.
/// </summary>
/// <remarks>
/// This is the difference from the retained inline-tagged JSON representation, whose values are
/// self-describing. Because the wire carries no discriminator, a value that does not match its
/// declared Shape is a decode failure, not a coercion.
/// </remarks>
public static class PortableValueCodec
{
    public static CborItem Encode(
        PortableShapeCatalog catalog,
        PortableShapeReference shape,
        PortableValue value)
    {
        ArgumentNullException.ThrowIfNull(catalog);
        ArgumentNullException.ThrowIfNull(value);
        var declaration = catalog.Shape(shape);
        switch (declaration.Kind)
        {
            case PortableShapeKind.Unit:
                return value is PortableUnitValue
                    ? CborNull.Instance
                    : throw Mismatch(shape, "unit");
            case PortableShapeKind.Scalar:
                return EncodeScalar(declaration, shape, value);
            case PortableShapeKind.Sequence:
            {
                var sequence = value as PortableSequenceValue ?? throw Mismatch(shape, "sequence");
                RequireBound(sequence.Items.Length, catalog.Limits.MaxSequenceItems, "sequence");
                return new CborArray(
                    [.. sequence.Items.Select(item => Encode(catalog, declaration.ItemShape!.Value, item))]);
            }
            case PortableShapeKind.Choice:
            {
                var choice = value as PortableChoiceValue ?? throw Mismatch(shape, "choice");
                var alternative = declaration.Alternatives
                    .FirstOrDefault(candidate => string.Equals(candidate.Name, choice.Alternative, StringComparison.Ordinal))
                    ?? throw PortableFaultException.InvalidPayload(
                        "unknown-alternative",
                        $"Shape {shape} declares no alternative '{choice.Alternative}'.");
                return CborArray.Of(
                    new CborText(choice.Alternative),
                    Encode(catalog, alternative.Shape, choice.Value));
            }
            case PortableShapeKind.Record:
                return EncodeRecord(catalog, declaration, shape, value);
            default:
                throw PortableFaultException.InvalidPayload("shape-kind", $"Shape {shape} has an unsupported kind.");
        }
    }

    public static PortableValue Decode(
        PortableShapeCatalog catalog,
        PortableShapeReference shape,
        CborItem item,
        IReadOnlyCollection<PortableFragmentReference>? operationFragments = null)
    {
        ArgumentNullException.ThrowIfNull(catalog);
        ArgumentNullException.ThrowIfNull(item);
        var declaration = catalog.Shape(shape);
        switch (declaration.Kind)
        {
            case PortableShapeKind.Unit:
                return item is CborNull
                    ? PortableUnitValue.Instance
                    : throw Mismatch(shape, "unit");
            case PortableShapeKind.Scalar:
                return DecodeScalar(declaration, shape, item);
            case PortableShapeKind.Sequence:
            {
                var array = item as CborArray ?? throw Mismatch(shape, "sequence");
                RequireBound(array.Items.Length, catalog.Limits.MaxSequenceItems, "sequence");
                return new PortableSequenceValue(
                    [.. array.Items.Select(element => Decode(catalog, declaration.ItemShape!.Value, element, operationFragments))]);
            }
            case PortableShapeKind.Choice:
            {
                if (item is not CborArray choice || choice.Items.Length != 2 || choice.Items[0] is not CborText name)
                {
                    throw Mismatch(shape, "choice");
                }

                var alternative = declaration.Alternatives
                    .FirstOrDefault(candidate => string.Equals(candidate.Name, name.Value, StringComparison.Ordinal))
                    ?? throw PortableFaultException.InvalidPayload(
                        "unknown-alternative",
                        $"Shape {shape} declares no alternative '{name.Value}'; the envelope kind was recognized, so this is a payload decision.");
                return new PortableChoiceValue(
                    name.Value,
                    Decode(catalog, alternative.Shape, choice.Items[1], operationFragments));
            }
            case PortableShapeKind.Record:
                return DecodeRecord(catalog, declaration, shape, item, operationFragments ?? []);
            default:
                throw PortableFaultException.InvalidPayload("shape-kind", $"Shape {shape} has an unsupported kind.");
        }
    }

    /// <summary>
    /// Validates an in-memory value against its declared Shape without producing bytes.
    /// </summary>
    /// <remarks>
    /// The fixed direct-call realization has no wire, so this is where it applies the same
    /// conformance rules the decoder applies. Encoding and decoding to check a direct call would
    /// manufacture a copy the realization does not make, and copy accounting is a reported fact.
    /// </remarks>
    public static void Validate(
        PortableShapeCatalog catalog,
        PortableShapeReference shape,
        PortableValue value,
        IReadOnlyCollection<PortableFragmentReference>? operationFragments = null)
    {
        ArgumentNullException.ThrowIfNull(catalog);
        ArgumentNullException.ThrowIfNull(value);
        var declaration = catalog.Shape(shape);
        var fragments = operationFragments ?? [];
        switch (declaration.Kind)
        {
            case PortableShapeKind.Unit:
                if (value is not PortableUnitValue)
                {
                    throw Mismatch(shape, "unit");
                }

                break;
            case PortableShapeKind.Scalar:
                _ = EncodeScalar(declaration, shape, value);
                break;
            case PortableShapeKind.Sequence:
            {
                var sequence = value as PortableSequenceValue ?? throw Mismatch(shape, "sequence");
                RequireBound(sequence.Items.Length, catalog.Limits.MaxSequenceItems, "sequence");
                foreach (var item in sequence.Items)
                {
                    Validate(catalog, declaration.ItemShape!.Value, item, fragments);
                }

                break;
            }
            case PortableShapeKind.Choice:
            {
                var choice = value as PortableChoiceValue ?? throw Mismatch(shape, "choice");
                var alternative = declaration.Alternatives
                    .FirstOrDefault(candidate => string.Equals(candidate.Name, choice.Alternative, StringComparison.Ordinal))
                    ?? throw PortableFaultException.InvalidPayload(
                        "unknown-alternative",
                        $"Shape {shape} declares no alternative '{choice.Alternative}'.");
                Validate(catalog, alternative.Shape, choice.Value, fragments);
                break;
            }
            case PortableShapeKind.Record:
                ValidateRecord(catalog, declaration, shape, value, fragments);
                break;
            default:
                throw PortableFaultException.InvalidPayload("shape-kind", $"Shape {shape} has an unsupported kind.");
        }
    }

    private static void ValidateRecord(
        PortableShapeCatalog catalog,
        PortableShapeDeclaration declaration,
        PortableShapeReference shape,
        PortableValue value,
        IReadOnlyCollection<PortableFragmentReference> operationFragments)
    {
        var record = value as PortableRecordValue ?? throw Mismatch(shape, "record");
        RequireBound(record.Fields.Count, catalog.Limits.MaxRecordFields, "record");
        RequireBound(record.Fragments.Count, catalog.Limits.MaxFragmentsPerRecord, "fragment map");

        var declared = declaration.Fields.ToImmutableDictionary(field => field.Name, StringComparer.Ordinal);
        foreach (var field in record.Fields)
        {
            if (!declared.TryGetValue(field.Key, out var fieldDeclaration))
            {
                throw PortableFaultException.InvalidPayload(
                    "undeclared-field",
                    $"Shape {shape} declares no field '{field.Key}'.");
            }

            Validate(catalog, fieldDeclaration.Shape, field.Value, operationFragments);
        }

        foreach (var field in declaration.Fields.Where(field => field.Required))
        {
            if (!record.Fields.ContainsKey(field.Name))
            {
                throw PortableFaultException.InvalidPayload(
                    "required-field-absent",
                    $"Shape {shape} requires field '{field.Name}'.");
            }
        }

        var permitted = operationFragments.ToImmutableHashSet();
        foreach (var fragment in record.Fragments)
        {
            if (declaration.FragmentPolicy == PortableFragmentPolicy.Closed && !permitted.Contains(fragment.Key))
            {
                throw PortableFaultException.InvalidPayload(
                    "closed-fragment-policy",
                    $"Shape {shape} is closed and rejects Fragment {fragment.Key}, which the negotiated Operation does not declare.");
            }

            if (!catalog.TryFragment(fragment.Key, out var fragmentDeclaration))
            {
                throw PortableFaultException.InvalidPayload(
                    "undeclared-fragment",
                    $"Fragment {fragment.Key} is not declared by the established contract, so its fields have no Shape.");
            }

            if (fragmentDeclaration.HostShape.Name != shape.Name || fragmentDeclaration.HostShape.Version > shape.Version)
            {
                throw PortableFaultException.InvalidPayload(
                    "fragment-host",
                    $"Fragment {fragment.Key} is not attachable to {shape}.");
            }

            var fragmentFields = fragmentDeclaration.Fields.ToImmutableDictionary(field => field.Name, StringComparer.Ordinal);
            foreach (var field in fragment.Value)
            {
                if (!fragmentFields.TryGetValue(field.Key, out var fieldDeclaration))
                {
                    throw PortableFaultException.InvalidPayload(
                        "undeclared-field",
                        $"Fragment {fragment.Key} declares no field '{field.Key}'.");
                }

                Validate(catalog, fieldDeclaration.Shape, field.Value, operationFragments);
            }

            foreach (var field in fragmentDeclaration.Fields.Where(field => field.Required))
            {
                if (!fragment.Value.ContainsKey(field.Name))
                {
                    throw PortableFaultException.InvalidPayload(
                        "required-field-absent",
                        $"Fragment {fragment.Key} requires field '{field.Name}'.");
                }
            }
        }
    }

    /// <summary>
    /// Projects a presented value onto the Shape version the consumer recognizes.
    /// </summary>
    /// <remarks>
    /// Projection discards unrecognized additive structure. It is therefore never applied to an
    /// authority or control position: discarding narrowing semantics from an authority value would
    /// widen authority.
    /// </remarks>
    public static PortableProjection Project(
        PortableShapeCatalog catalog,
        PortableValue value,
        PortableShapeReference presented,
        PortableShapeReference accepted,
        IReadOnlyCollection<PortableFragmentReference> operationFragments)
    {
        ArgumentNullException.ThrowIfNull(catalog);
        ArgumentNullException.ThrowIfNull(value);
        ArgumentNullException.ThrowIfNull(operationFragments);

        if (presented == accepted)
        {
            return new PortableProjection(value, []);
        }

        if (!catalog.IsAdditiveOver(presented, accepted))
        {
            throw PortableFaultException.InvalidPayload(
                "non-additive-projection",
                $"Shape {presented} does not differ from {accepted} additively, so it cannot be projected.");
        }

        var acceptedShape = catalog.Shape(accepted);
        if (acceptedShape.Kind != PortableShapeKind.Record || value is not PortableRecordValue record)
        {
            return new PortableProjection(value, [$"projected:{presented}->{accepted}"]);
        }

        var declaredNames = acceptedShape.Fields.Select(field => field.Name).ToImmutableHashSet(StringComparer.Ordinal);
        var retained = record.Fields
            .Where(field => declaredNames.Contains(field.Key))
            .ToImmutableDictionary(field => field.Key, field => field.Value, StringComparer.Ordinal);
        var dropped = record.Fields.Keys
            .Where(name => !declaredNames.Contains(name))
            .OrderBy(name => name, StringComparer.Ordinal)
            .ToImmutableArray();

        var obligations = ImmutableArray.CreateBuilder<string>();
        obligations.Add($"projected:{presented}->{accepted}");
        foreach (var name in dropped)
        {
            obligations.Add($"field-projected:{name}");
        }

        return new PortableProjection(record with { Fields = retained }, obligations.ToImmutable());
    }

    // ---------------------------------------------------------------------

    private static CborItem EncodeScalar(
        PortableShapeDeclaration declaration,
        PortableShapeReference shape,
        PortableValue value) => (declaration.Scalar, value) switch
    {
        (PortableScalarKind.Text, PortableTextValue text) => new CborText(text.Value),
        (PortableScalarKind.Boolean, PortableBooleanValue boolean) => new CborBoolean(boolean.Value),
        (PortableScalarKind.Signed64, PortableIntegerValue integer) => new CborInteger(integer.Value),
        (PortableScalarKind.Decimal, PortableDecimalValue number) => CborDecimal.Normalize(number.Exponent, number.Mantissa),
        (PortableScalarKind.Bytes, PortableBytesValue bytes) => new CborBytes(bytes.Value),
        _ => throw Mismatch(shape, declaration.Scalar?.Token() ?? "scalar")
    };

    private static PortableValue DecodeScalar(
        PortableShapeDeclaration declaration,
        PortableShapeReference shape,
        CborItem item) => (declaration.Scalar, item) switch
    {
        (PortableScalarKind.Text, CborText text) => new PortableTextValue(text.Value),
        (PortableScalarKind.Boolean, CborBoolean boolean) => new PortableBooleanValue(boolean.Value),
        (PortableScalarKind.Signed64, CborInteger integer) => new PortableIntegerValue(integer.Value),
        (PortableScalarKind.Decimal, CborDecimal number) => new PortableDecimalValue(number.Exponent, number.Mantissa),
        (PortableScalarKind.Bytes, CborBytes bytes) => new PortableBytesValue(bytes.Value),
        _ => throw Mismatch(shape, declaration.Scalar?.Token() ?? "scalar")
    };

    private static CborItem EncodeRecord(
        PortableShapeCatalog catalog,
        PortableShapeDeclaration declaration,
        PortableShapeReference shape,
        PortableValue value)
    {
        var record = value as PortableRecordValue ?? throw Mismatch(shape, "record");
        RequireBound(record.Fields.Count, catalog.Limits.MaxRecordFields, "record");
        RequireBound(record.Fragments.Count, catalog.Limits.MaxFragmentsPerRecord, "fragment map");

        var fields = new List<(string, CborItem)>();
        foreach (var field in declaration.Fields)
        {
            if (!record.Fields.TryGetValue(field.Name, out var fieldValue))
            {
                if (field.Required)
                {
                    throw PortableFaultException.InvalidPayload(
                        "required-field-absent",
                        $"Shape {shape} requires field '{field.Name}'.");
                }

                // An optional field carrying no value is omitted rather than encoded as null;
                // omission is what makes additive projection possible.
                continue;
            }

            fields.Add((field.Name, Encode(catalog, field.Shape, fieldValue)));
        }

        var declaredNames = declaration.Fields.Select(field => field.Name).ToImmutableHashSet(StringComparer.Ordinal);
        var undeclared = record.Fields.Keys.FirstOrDefault(name => !declaredNames.Contains(name));
        if (undeclared is not null)
        {
            throw PortableFaultException.InvalidPayload(
                "undeclared-field",
                $"Shape {shape} declares no field '{undeclared}'.");
        }

        var fragments = new List<(string, CborItem)>();
        foreach (var fragment in record.Fragments)
        {
            var fragmentDeclaration = catalog.Fragment(fragment.Key);
            var fragmentFields = new List<(string, CborItem)>();
            foreach (var field in fragmentDeclaration.Fields)
            {
                if (!fragment.Value.TryGetValue(field.Name, out var fieldValue))
                {
                    if (field.Required)
                    {
                        throw PortableFaultException.InvalidPayload(
                            "required-field-absent",
                            $"Fragment {fragment.Key} requires field '{field.Name}'.");
                    }

                    continue;
                }

                fragmentFields.Add((field.Name, Encode(catalog, field.Shape, fieldValue)));
            }

            fragments.Add((fragment.Key.CanonicalText, CborMap.Of(fragmentFields)));
        }

        return CborArray.Of(CborMap.Of(fields), CborMap.Of(fragments));
    }

    private static PortableValue DecodeRecord(
        PortableShapeCatalog catalog,
        PortableShapeDeclaration declaration,
        PortableShapeReference shape,
        CborItem item,
        IReadOnlyCollection<PortableFragmentReference> operationFragments)
    {
        if (item is not CborArray array || array.Items.Length != 2 ||
            array.Items[0] is not CborMap fieldMap || array.Items[1] is not CborMap fragmentMap)
        {
            throw Mismatch(shape, "record");
        }

        RequireBound(fieldMap.Entries.Length, catalog.Limits.MaxRecordFields, "record");
        RequireBound(fragmentMap.Entries.Length, catalog.Limits.MaxFragmentsPerRecord, "fragment map");

        var declared = declaration.Fields.ToImmutableDictionary(field => field.Name, StringComparer.Ordinal);
        var fields = ImmutableDictionary.CreateBuilder<string, PortableValue>(StringComparer.Ordinal);
        foreach (var entry in fieldMap.Entries)
        {
            if (!declared.TryGetValue(entry.Key, out var field))
            {
                throw PortableFaultException.InvalidPayload(
                    "undeclared-field",
                    $"Shape {shape} declares no field '{entry.Key}'.");
            }

            fields[entry.Key] = Decode(catalog, field.Shape, entry.Value, operationFragments);
        }

        foreach (var field in declaration.Fields.Where(field => field.Required))
        {
            if (!fields.ContainsKey(field.Name))
            {
                throw PortableFaultException.InvalidPayload(
                    "required-field-absent",
                    $"Shape {shape} requires field '{field.Name}'.");
            }
        }

        var permitted = operationFragments.ToImmutableHashSet();
        var fragments = ImmutableDictionary.CreateBuilder<PortableFragmentReference, ImmutableDictionary<string, PortableValue>>();
        foreach (var entry in fragmentMap.Entries)
        {
            var reference = PortableFragmentReference.ParseText(entry.Key);
            if (declaration.FragmentPolicy == PortableFragmentPolicy.Closed && !permitted.Contains(reference))
            {
                throw PortableFaultException.InvalidPayload(
                    "closed-fragment-policy",
                    $"Shape {shape} is closed and rejects Fragment {reference}, which the negotiated Operation does not declare.");
            }

            if (!catalog.TryFragment(reference, out var fragmentDeclaration))
            {
                throw PortableFaultException.InvalidPayload(
                    "undeclared-fragment",
                    $"Fragment {reference} is not declared by the established contract, so its fields have no Shape.");
            }

            if (fragmentDeclaration.HostShape.Name != shape.Name || fragmentDeclaration.HostShape.Version > shape.Version)
            {
                throw PortableFaultException.InvalidPayload(
                    "fragment-host",
                    $"Fragment {reference} is not attachable to {shape}.");
            }

            fragments[reference] = DecodeFragmentFields(
                catalog,
                fragmentDeclaration,
                CborAccess.RequireMap(entry.Value, "fragment"),
                operationFragments);
        }

        return new PortableRecordValue(fields.ToImmutable(), fragments.ToImmutable());
    }

    private static ImmutableDictionary<string, PortableValue> DecodeFragmentFields(
        PortableShapeCatalog catalog,
        PortableFragmentDeclaration fragment,
        CborMap map,
        IReadOnlyCollection<PortableFragmentReference> operationFragments)
    {
        var declared = fragment.Fields.ToImmutableDictionary(field => field.Name, StringComparer.Ordinal);
        var fields = ImmutableDictionary.CreateBuilder<string, PortableValue>(StringComparer.Ordinal);
        foreach (var entry in map.Entries)
        {
            if (!declared.TryGetValue(entry.Key, out var field))
            {
                throw PortableFaultException.InvalidPayload(
                    "undeclared-field",
                    $"Fragment {fragment.Reference} declares no field '{entry.Key}'.");
            }

            fields[entry.Key] = Decode(catalog, field.Shape, entry.Value, operationFragments);
        }

        foreach (var field in fragment.Fields.Where(field => field.Required))
        {
            if (!fields.ContainsKey(field.Name))
            {
                throw PortableFaultException.InvalidPayload(
                    "required-field-absent",
                    $"Fragment {fragment.Reference} requires field '{field.Name}'.");
            }
        }

        return fields.ToImmutable();
    }

    private static void RequireBound(int count, int bound, string what)
    {
        if (count > bound)
        {
            throw PortableFaultException.LimitExceeded(
                $"{what}-bound",
                $"A {what} of {count} exceeds the declared bound of {bound}.");
        }
    }

    private static PortableFaultException Mismatch(PortableShapeReference shape, string expected) =>
        PortableFaultException.InvalidPayload(
            "shape-mismatch",
            $"Shape {shape} requires a {expected} value.");
}

File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableValueCodec.fs
namespace Brontide.Minimal.Binding.Portable

/// The schema-guided value codec: the negotiated Shape determines the type of every position, so
/// values carry no kind discriminator on the wire.
///
/// This is the difference from the retained inline-tagged JSON representation, whose values are
/// self-describing. Because the wire carries no discriminator, a value that does not match its
/// declared Shape is a refusal rather than a coercion.
[<RequireQualifiedAccess>]
module PortableValueCodec =

    let private mismatch reference expected =
        invalidPayload "shape-mismatch" $"Shape {PortableShapeRef.text reference} requires a {expected} value."

    let private requireBound count bound what =
        if count > bound then
            limitExceeded $"{what}-bound" $"A {what} of {count} exceeds the declared bound of {bound}."
        else
            Ok()

    let private encodeScalar scalar reference value =
        match scalar, value with
        | PortableScalar.Text, PortableText text -> Ok(CborText text)
        | PortableScalar.Boolean, PortableBoolean flag -> Ok(CborBoolean flag)
        | PortableScalar.Signed64, PortableInteger number -> Ok(CborInteger number)
        | PortableScalar.Decimal, PortableDecimalValue(exponent, mantissa) ->
            CborDecimal.normalize exponent mantissa
            |> Result.map (fun (exponent, mantissa) -> CborDecimal(exponent, mantissa))
        | PortableScalar.Bytes, PortableBytesValue bytes -> Ok(CborBytes bytes)
        | _ -> mismatch reference (PortableScalar.token scalar)

    let private decodeScalar scalar reference item =
        match scalar, item with
        | PortableScalar.Text, CborText text -> Ok(PortableText text)
        | PortableScalar.Boolean, CborBoolean flag -> Ok(PortableBoolean flag)
        | PortableScalar.Signed64, CborInteger number -> Ok(PortableInteger number)
        | PortableScalar.Decimal, CborDecimal(exponent, mantissa) -> Ok(PortableDecimalValue(exponent, mantissa))
        | PortableScalar.Bytes, CborBytes bytes -> Ok(PortableBytesValue bytes)
        | _ -> mismatch reference (PortableScalar.token scalar)

    let private alternativeOf alternatives reference name =
        match alternatives |> List.tryFind (fun (candidate: AlternativeDeclaration) -> candidate.Name = name) with
        | Some alternative -> Ok alternative
        | None ->
            invalidPayload
                "unknown-alternative"
                $"Shape {PortableShapeRef.text reference} declares no alternative '{name}'; the envelope kind was recognized, so this is a payload decision."

    /// Refuses a Fragment the negotiated Operation does not declare on a closed Shape, an
    /// undeclared Fragment, and a Fragment whose host Shape is not this one.
    let private admitFragment catalog (shapeRef: PortableShapeRef) policy operationFragments reference =
        portable {
            do!
                ensure (policy = FragmentPolicy.Open || List.contains reference operationFragments) (fun () ->
                    invalidPayload
                        "closed-fragment-policy"
                        $"Shape {PortableShapeRef.text shapeRef} is closed and refuses Fragment {PortableFragmentRef.text reference}, which the negotiated Operation does not declare.")

            let! declaration =
                match ShapeCatalog.tryFragment reference catalog with
                | Some declaration -> Ok declaration
                | None ->
                    invalidPayload
                        "undeclared-fragment"
                        $"Fragment {PortableFragmentRef.text reference} is not declared by the established contract, so its fields have no Shape."

            do!
                ensure
                    (PortableShapeRef.name declaration.HostShape = PortableShapeRef.name shapeRef
                     && PortableShapeRef.version declaration.HostShape <= PortableShapeRef.version shapeRef)
                    (fun () ->
                        invalidPayload
                            "fragment-host"
                            $"Fragment {PortableFragmentRef.text reference} is not attachable to {PortableShapeRef.text shapeRef}.")

            return declaration
        }

    let private requireRequiredFields owner (declared: FieldDeclaration list) (present: string -> bool) =
        declared
        |> iterate (fun field ->
            if field.Required && not (present field.Name) then
                invalidPayload "required-field-absent" $"{owner} requires field '{field.Name}'."
            else
                Ok())

    // -- encoding -----------------------------------------------------------

    let rec encode catalog reference value : PortableResult<CborItem> =
        portable {
            let! declaration = ShapeCatalog.shape reference catalog

            match declaration.Body with
            | UnitBody ->
                match value with
                | PortableUnit -> return CborNull
                | _ -> return! mismatch reference "unit"
            | ScalarBody scalar -> return! encodeScalar scalar reference value
            | SequenceBody item ->
                match value with
                | PortableSequence items ->
                    do! requireBound (List.length items) (ShapeCatalog.limits catalog).MaxSequenceItems "sequence"
                    let! encoded = items |> traverse (encode catalog item)
                    return CborArray encoded
                | _ -> return! mismatch reference "sequence"
            | ChoiceBody alternatives ->
                match value with
                | PortableChoice(name, inner) ->
                    let! alternative = alternativeOf alternatives reference name
                    let! encoded = encode catalog alternative.Shape inner
                    return CborArray [ CborText name; encoded ]
                | _ -> return! mismatch reference "choice"
            | RecordBody(_, declaredFields) -> return! encodeRecord catalog reference declaredFields value
        }

    and private encodeRecord catalog reference declaredFields value =
        match value with
        | PortableRecord(fields, fragments) ->
            portable {
                let bounds = ShapeCatalog.limits catalog
                do! requireBound (Map.count fields) bounds.MaxRecordFields "record"
                do! requireBound (Map.count fragments) bounds.MaxFragmentsPerRecord "fragment map"

                let declaredNames = declaredFields |> List.map (fun field -> field.Name) |> Set.ofList

                do!
                    fields
                    |> Map.toList
                    |> iterate (fun (name, _) ->
                        if Set.contains name declaredNames then
                            Ok()
                        else
                            invalidPayload
                                "undeclared-field"
                                $"Shape {PortableShapeRef.text reference} declares no field '{name}'.")

                // An optional field carrying no value is omitted rather than encoded as null;
                // omission is what makes additive projection possible.
                let! encodedFields =
                    declaredFields
                    |> traverse (fun field ->
                        match Map.tryFind field.Name fields with
                        | Some fieldValue -> encode catalog field.Shape fieldValue |> Result.map (fun item -> [ field.Name, item ])
                        | None when field.Required ->
                            invalidPayload
                                "required-field-absent"
                                $"Shape {PortableShapeRef.text reference} requires field '{field.Name}'."
                        | None -> Ok [])

                let! encodedFragments =
                    fragments
                    |> Map.toList
                    |> traverse (fun (fragmentRef, fragmentFields) ->
                        portable {
                            let! declaration = ShapeCatalog.fragment fragmentRef catalog

                            let! encoded =
                                declaration.Fields
                                |> traverse (fun field ->
                                    match Map.tryFind field.Name fragmentFields with
                                    | Some fieldValue ->
                                        encode catalog field.Shape fieldValue
                                        |> Result.map (fun item -> [ field.Name, item ])
                                    | None when field.Required ->
                                        invalidPayload
                                            "required-field-absent"
                                            $"Fragment {PortableFragmentRef.text fragmentRef} requires field '{field.Name}'."
                                    | None -> Ok [])

                            return PortableFragmentRef.text fragmentRef, CborMap(List.concat encoded)
                        })

                return CborArray [ CborMap(List.concat encodedFields); CborMap encodedFragments ]
            }
        | _ -> mismatch reference "record"

    // -- decoding -----------------------------------------------------------

    let rec decode catalog reference operationFragments item : PortableResult<PortableValue> =
        portable {
            let! declaration = ShapeCatalog.shape reference catalog

            match declaration.Body with
            | UnitBody ->
                match item with
                | CborNull -> return PortableUnit
                | _ -> return! mismatch reference "unit"
            | ScalarBody scalar -> return! decodeScalar scalar reference item
            | SequenceBody itemShape ->
                match item with
                | CborArray items ->
                    do! requireBound (List.length items) (ShapeCatalog.limits catalog).MaxSequenceItems "sequence"
                    let! decoded = items |> traverse (decode catalog itemShape operationFragments)
                    return PortableSequence decoded
                | _ -> return! mismatch reference "sequence"
            | ChoiceBody alternatives ->
                match item with
                | CborArray [ CborText name; body ] ->
                    let! alternative = alternativeOf alternatives reference name
                    let! decoded = decode catalog alternative.Shape operationFragments body
                    return PortableChoice(name, decoded)
                | _ -> return! mismatch reference "choice"
            | RecordBody(policy, declaredFields) ->
                return! decodeRecord catalog reference policy declaredFields operationFragments item
        }

    and private decodeRecord catalog reference policy declaredFields operationFragments item =
        match item with
        | CborArray [ CborMap fieldEntries; CborMap fragmentEntries ] ->
            portable {
                let bounds = ShapeCatalog.limits catalog
                do! requireBound (List.length fieldEntries) bounds.MaxRecordFields "record"
                do! requireBound (List.length fragmentEntries) bounds.MaxFragmentsPerRecord "fragment map"

                let declaredByName =
                    declaredFields |> List.map (fun field -> field.Name, field) |> Map.ofList

                let! fields =
                    fieldEntries
                    |> traverse (fun (name, value) ->
                        match Map.tryFind name declaredByName with
                        | Some field ->
                            decode catalog field.Shape operationFragments value
                            |> Result.map (fun decoded -> name, decoded)
                        | None ->
                            invalidPayload
                                "undeclared-field"
                                $"Shape {PortableShapeRef.text reference} declares no field '{name}'.")

                let fieldMap = Map.ofList fields

                do!
                    requireRequiredFields
                        $"Shape {PortableShapeRef.text reference}"
                        declaredFields
                        (fun name -> Map.containsKey name fieldMap)

                let! fragments =
                    fragmentEntries
                    |> traverse (fun (key, value) ->
                        portable {
                            let! fragmentRef = PortableFragmentRef.tryParseText key
                            let! declaration = admitFragment catalog reference policy operationFragments fragmentRef
                            let! entries = CborAccess.requireMap value "fragment"
                            let! decoded = decodeFragmentFields catalog declaration operationFragments entries
                            return fragmentRef, decoded
                        })

                return PortableRecord(fieldMap, Map.ofList fragments)
            }
        | _ -> mismatch reference "record"

    and private decodeFragmentFields catalog (declaration: FragmentDeclaration) operationFragments entries =
        portable {
            let declaredByName =
                declaration.Fields |> List.map (fun field -> field.Name, field) |> Map.ofList

            let! fields =
                entries
                |> traverse (fun (name, value) ->
                    match Map.tryFind name declaredByName with
                    | Some field ->
                        decode catalog field.Shape operationFragments value
                        |> Result.map (fun decoded -> name, decoded)
                    | None ->
                        invalidPayload
                            "undeclared-field"
                            $"Fragment {PortableFragmentRef.text declaration.Reference} declares no field '{name}'.")

            let fieldMap = Map.ofList fields

            do!
                requireRequiredFields
                    $"Fragment {PortableFragmentRef.text declaration.Reference}"
                    declaration.Fields
                    (fun name -> Map.containsKey name fieldMap)

            return fieldMap
        }

    // -- validation ---------------------------------------------------------

    /// Validates an in-memory value against its declared Shape without producing bytes.
    ///
    /// The fixed direct-call realization has no wire, so this is where it applies the same
    /// conformance rules the decoder applies. Encoding and decoding to check a direct call would
    /// manufacture a copy the realization does not make, and copy accounting is a reported fact.
    let rec validate catalog reference operationFragments value : PortableResult<unit> =
        portable {
            let! declaration = ShapeCatalog.shape reference catalog

            match declaration.Body with
            | UnitBody ->
                match value with
                | PortableUnit -> ()
                | _ -> return! mismatch reference "unit"
            | ScalarBody scalar ->
                let! _ = encodeScalar scalar reference value
                ()
            | SequenceBody itemShape ->
                match value with
                | PortableSequence items ->
                    do! requireBound (List.length items) (ShapeCatalog.limits catalog).MaxSequenceItems "sequence"
                    do! items |> iterate (validate catalog itemShape operationFragments)
                | _ -> return! mismatch reference "sequence"
            | ChoiceBody alternatives ->
                match value with
                | PortableChoice(name, inner) ->
                    let! alternative = alternativeOf alternatives reference name
                    do! validate catalog alternative.Shape operationFragments inner
                | _ -> return! mismatch reference "choice"
            | RecordBody(policy, declaredFields) ->
                match value with
                | PortableRecord(fields, fragments) ->
                    let bounds = ShapeCatalog.limits catalog
                    do! requireBound (Map.count fields) bounds.MaxRecordFields "record"
                    do! requireBound (Map.count fragments) bounds.MaxFragmentsPerRecord "fragment map"

                    let declaredByName =
                        declaredFields |> List.map (fun field -> field.Name, field) |> Map.ofList

                    do!
                        fields
                        |> Map.toList
                        |> iterate (fun (name, fieldValue) ->
                            match Map.tryFind name declaredByName with
                            | Some field -> validate catalog field.Shape operationFragments fieldValue
                            | None ->
                                invalidPayload
                                    "undeclared-field"
                                    $"Shape {PortableShapeRef.text reference} declares no field '{name}'.")

                    do!
                        requireRequiredFields
                            $"Shape {PortableShapeRef.text reference}"
                            declaredFields
                            (fun name -> Map.containsKey name fields)

                    do!
                        fragments
                        |> Map.toList
                        |> iterate (fun (fragmentRef, fragmentFields) ->
                            portable {
                                let! declaration = admitFragment catalog reference policy operationFragments fragmentRef

                                let declaredFragmentFields =
                                    declaration.Fields |> List.map (fun field -> field.Name, field) |> Map.ofList

                                do!
                                    fragmentFields
                                    |> Map.toList
                                    |> iterate (fun (name, fieldValue) ->
                                        match Map.tryFind name declaredFragmentFields with
                                        | Some field -> validate catalog field.Shape operationFragments fieldValue
                                        | None ->
                                            invalidPayload
                                                "undeclared-field"
                                                $"Fragment {PortableFragmentRef.text fragmentRef} declares no field '{name}'.")

                                do!
                                    requireRequiredFields
                                        $"Fragment {PortableFragmentRef.text fragmentRef}"
                                        declaration.Fields
                                        (fun name -> Map.containsKey name fragmentFields)
                            })
                | _ -> return! mismatch reference "record"
        }

    // -- projection ---------------------------------------------------------

    /// Projects a presented value onto the Shape version the consumer recognizes, reporting the
    /// mapping obligations the projection discharged.
    ///
    /// Projection discards unrecognized additive structure. It is therefore never applied to an
    /// authority or control position: discarding narrowing semantics would widen authority.
    let project catalog presented accepted value : PortableResult<PortableValue * string list> =
        if presented = accepted then
            Ok(value, [])
        elif not (ShapeCatalog.isAdditiveOver presented accepted catalog) then
            invalidPayload
                "non-additive-projection"
                $"Shape {PortableShapeRef.text presented} does not differ from {PortableShapeRef.text accepted} additively, so it cannot be projected."
        else
            let obligation =
                $"projected:{PortableShapeRef.text presented}->{PortableShapeRef.text accepted}"

            ShapeCatalog.shape accepted catalog
            |> Result.map (fun declaration ->
                match declaration.Body, value with
                | RecordBody(_, declaredFields), PortableRecord(fields, fragments) ->
                    let declaredNames = declaredFields |> List.map (fun field -> field.Name) |> Set.ofList

                    let retained =
                        fields |> Map.filter (fun name _ -> Set.contains name declaredNames)

                    let dropped =
                        fields
                        |> Map.toList
                        |> List.map fst
                        |> List.filter (fun name -> not (Set.contains name declaredNames))
                        |> List.sort

                    PortableRecord(retained, fragments), obligation :: (dropped |> List.map (fun name -> $"field-projected:{name}"))
                | _ -> value, [ obligation ])
